common/lib/src/realm_types.dart

```diff
@@ -104,35 +103,60 @@ class RealmStateError extends StateError implements RealmError {
 class Decimal128 {} // TODO Support decimal128 datatype https://github.com/realm/realm-dart/issues/725
 
 /// @nodoc
-class RealmObjectBaseMarker {}
-
-/// @nodoc
-class RealmObjectMarker extends RealmObjectBaseMarker {}
-
-/// @nodoc
-class EmbeddedObjectMarker extends RealmObjectBaseMarker {}
+abstract class RealmObjectBaseMarker {}
 
 // Union type
 /// @nodoc
-class RealmAny {
-  final dynamic value;
+class RealmValue {
+  final Object? value;
+  Type get type => value.runtimeType;
   T as<T>() => value as T; // better for code completion
 
   // This is private, so user cannot accidentally construct an invalid instance
-  const RealmAny._(this.value);
-
-  const RealmAny.bool(bool b) : this._(b);
-  const RealmAny.string(String text) : this._(text);
-  const RealmAny.int(int i) : this._(i);
-  const RealmAny.float(Float f) : this._(f);
-  const RealmAny.double(double d) : this._(d);
-  const RealmAny.uint8List(Uint8List data) : this._(data);
+  const RealmValue._(this.value);
+
+  const RealmValue.nullValue() : this._(null);
+  const RealmValue.bool(bool b) : this._(b);
+  const RealmValue.string(String text) : this._(text);
+  const RealmValue.int(int i) : this._(i);
+  const RealmValue.double(double d) : this._(d);
+  // const RealmValue.uint8List(Uint8List data) : this._(data);
   // TODO: RealmObjectMarker introduced to avoid dependency inversion. It would be better if we could use RealmObject directly. https://github.com/realm/realm-dart/issues/701
-  const RealmAny.realmObject(RealmObjectBaseMarker o) : this._(o);
-  const RealmAny.dateTime(DateTime timestamp) : this._(timestamp);
-  const RealmAny.objectId(ObjectId id) : this._(id);
-  const RealmAny.decimal128(Decimal128 decimal) : this._(decimal);
-  const RealmAny.uuid(Uuid uuid) : this._(uuid);
+  const RealmValue.realmObject(RealmObjectBaseMarker o) : this._(o);
+  const RealmValue.dateTime(DateTime timestamp) : this._(timestamp);
+  const RealmValue.objectId(ObjectId id) : this._(id);
+  // const RealmValue.decimal128(Decimal128 decimal) : this._(decimal); // not supported yet
+  const RealmValue.uuid(Uuid uuid) : this._(uuid);
+
+  factory RealmValue.from(Object? o) {
+    if (o == null ||
+        o is bool ||
+        o is String ||
+        o is int ||
+        o is Float ||
+        o is double ||
+        o is RealmObjectBaseMarker ||
+        o is DateTime ||
+        o is ObjectId ||
+        // o is Decimal128 || // not supported yet
+        o is Uuid) {
+      return RealmValue._(o);
+    } else {
+      throw ArgumentError.value(o, 'o', 'Unsupported type');
+    }
+  }
+
+  @override
+  operator ==(Object? other) {
+    if (other is RealmValue) {
+      return value == other.value;
+    } else {
+      return value == other;
```

Comment:
why not return `false` here. In what case this can be `true`

Reply:
This allows you to write `RealmValue.int(1) == 1` and get `true`

Reply:
got it